File: ch6/clip3/boj17225/main.py
b_time, r_time, n = map(int, input().split())
blue_queue = []
red_queue = []

for _ in range(n):
    time, color, cnt = input().split()
    time = int(time)
    cnt = int(cnt)

    if color == 'B':
        if blue_queue and blue_queue[-1] + b_time > time:
            time = blue_queue[-1] + b_time
        for _ in range(cnt):
            blue_queue.append(time)
            time += b_time
    elif color == 'R':
        if red_queue and red_queue[-1] + r_time > time:
            time = red_queue[-1] + r_time
        for _ in range(cnt):
            red_queue.append(time)
            time += r_time

blue_order = []
red_order = []
order_number = 0

while blue_queue and red_queue:
    if blue_queue[0] <= red_queue[0]:
        blue_order.append(order_number + 1)
        order_number += 1
        blue_queue.pop(0)
    else:
        red_order.append(order_number + 1)
        order_number += 1
        red_queue.pop(0)

while blue_queue:
    blue_order.append(order_number + 1)
    order_number += 1
    blue_queue.pop(0)

while red_queue:
    red_order.append(order_number + 1)
    order_number += 1
    red_queue.pop(0)

print(len(blue_order))
print(" ".join(map(str, blue_order)))
print(len(red_order))
print(" ".join(map(str, red_order)))

File: ch6/clip3/boj17225/Main.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int bTime = sc.nextInt();
        int rTime = sc.nextInt();
        int n = sc.nextInt();
        Deque<Integer> blueQueue = new LinkedList<>();
        Deque<Integer> redQueue = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            int time = sc.nextInt();
            char color = sc.next().charAt(0);
            int cnt = sc.nextInt();

            switch (color) {
                case 'B' -> {
                    if (!blueQueue.isEmpty() && blueQueue.getLast() + bTime > time) {
                        time = blueQueue.getLast() + bTime;
                    }
                    for (int j = 0; j < cnt; j++) {
                        blueQueue.offer(time);
                        time += bTime;
                    }
                }
                case 'R' -> {
                    if (!redQueue.isEmpty() && redQueue.getLast() + rTime > time) {
                        time = redQueue.getLast() + rTime;
                    }
                    for (int j = 0; j < cnt; j++) {
                        redQueue.offer(time);
                        time += rTime;
                    }
                }
            }
        }
        List<Integer> blueOrder = new LinkedList<>();
        List<Integer> redOrder = new LinkedList<>();
        int orderNumber = 0;
        while (!blueQueue.isEmpty() && !redQueue.isEmpty()) {
            if(blueQueue.peek() <= redQueue.peek()) {
                blueOrder.add(++orderNumber);
                blueQueue.poll();
            }
            else {
                redOrder.add(++orderNumber);
                redQueue.poll();
            }
        }

        while(!blueQueue.isEmpty()) {
            blueOrder.add(++orderNumber);
            blueQueue.poll();
        }

        while(!redQueue.isEmpty()) {
            redOrder.add(++orderNumber);
            redQueue.poll();
        }


        System.out.println(blueOrder.size());
        for (Integer integer : blueOrder) {
            System.out.print(integer + " ");
        }
        System.out.println();
        System.out.println(redOrder.size());
        for (Integer integer : redOrder) {
            System.out.print(integer + " ");
        }
    }
}
